src/target/adiv5_swd.c

```diff
@@ -53,10 +53,66 @@ uint8_t make_packet_request(uint8_t RnW, uint16_t addr)
 
 /* Provide bare DP access functions without timeout and exception */
 
-static void swd_line_reset(void)
+static void swd_line_reset_sequence(const bool idle_cycles)
 {
-	swd_proc.seq_out(0xffffffffU, 32U);
-	swd_proc.seq_out(0x0fffffffU, 32U);
+	/* 
+	 * A line reset is achieved by holding the SWDIOTMS HIGH for at least 50 SWCLKTCK cycles, followed by at least two idle cycles
+	 * Note: in some non-conformant devices (STM32) at least 51 HIGH cycles and/or 3/4 idle cycles are required
+	 *
+	 * for robustness, we use 60 HIGH cycles and 4 idle cycles
+	 */
+	swd_proc.seq_out(0xffffffffU, 32U);                     /* 32 cycles HIGH */
+	swd_proc.seq_out(0x0fffffffU, idle_cycles ? 32U : 28U); /* 28 cycles HIGH + 4 idle cycles if idle is requested */
+}
+
+/* Switch out of dormant state into SWD */
+static void dormant_state_to_swd_sequence()
+{
+	/*
+	 * ARM Debug Interface Architecture Specification, ADIv5.0 to ADIv5.2. ARM IHI 0031C
+	 * §5.3.4 Switching out of Dormant state
+	 */
+
+	DEBUG_INFO("Switching out of dormant state into SWD\n");
+
+	/* Send at least 8 SWCLKTCK cycles with SWDIOTMS HIGH */
+	swd_line_reset_sequence(false);
+	/* Send the 128-bit Selection Alert sequence on SWDIOTMS */
+	swd_proc.seq_out(ADIV5_SELECTION_ALERT_SEQUENCE_0, 32U);
+	swd_proc.seq_out(ADIV5_SELECTION_ALERT_SEQUENCE_1, 32U);
+	swd_proc.seq_out(ADIV5_SELECTION_ALERT_SEQUENCE_2, 32U);
+	swd_proc.seq_out(ADIV5_SELECTION_ALERT_SEQUENCE_3, 32U);
+	/* Send 4 SWCLKTCK cycles with SWDIOTMS LOW */
+	/* Send the required activation code sequence on SWDIOTMS */
+	/* We combine the last two actions in a single seq_out as an optimization */
+	swd_proc.seq_out(ADIV5_ACTIVATION_CODE_ARM_SWD_DP << 4U, 8U + 4U);
```

Comment:
Please make this a proper block comment, and note how many cycles the activation code requires in the comment so the number passed can be simplified to just `12U` rather than relying on constant folding

Reply:
The reason they were separated was because I wanted to clearly show they were individual steps (as they are listed on the spec), that we combine into a single sequence. That said, it wasn't really that obvious or useful.